cli/internal/core/engine.go

```diff
@@ -385,11 +410,76 @@ func (e *Engine) ValidatePersistentDependencies(graph *graph.CompleteGraph) erro
 	return validationError
 }
 
-// GetResolvedTaskDefinition returns a "resolved" TaskDefinition.
-// Today, it just looks up the task from the root Pipeline, but in the future
-// we will compose the TaskDefinition from workspaces using the `extends` key.
-func (e *Engine) GetResolvedTaskDefinition(rootPipeline *fs.Pipeline, taskName string, taskID string) (*fs.TaskDefinition, error) {
-	return rootPipeline.GetTask(taskID, taskName)
+// getTaskDefinitionChain gets a set of TaskDefinitions that apply to the taskID.
+// These definitions should be merged by the consumer.
+// TODO(mehulkar): Can we use getPipelineFromWorkspace() here instead of reading configs again and passing around rootPipeline as an arg?
+func (e *Engine) getTaskDefinitionChain(rootPipeline *fs.Pipeline, pkg *fs.PackageJSON, taskID string, taskName string) ([]fs.BookkeepingTaskDefinition, error) {
+	// Start a list of TaskDefinitions we've found for this TaskID
+	taskDefinitions := []fs.BookkeepingTaskDefinition{}
+
+	// Look for the taskDefinition in the root pipeline.
+	// We'll wait to throw errors until the end
+	if rootTaskDefinition, err := rootPipeline.GetTask(taskID, taskName); err == nil {
+		taskDefinitions = append(taskDefinitions, *rootTaskDefinition)
+	}
+
+	// If the taskID is a root task (e.g. //#build), we don't need to look
+	// for a workspace task, since these can only be defined in the root turbo.json.
+	taskIDPackage, _ := util.GetPackageTaskFromId(taskID)
+	if taskIDPackage != util.RootPkgName && taskIDPackage != ROOT_NODE_NAME {
+		// Look up task definition in turbo.json in the workspace directory
+		workspaceConfigPath := pkg.Dir.Join("turbo.json").RestoreAnchor(e.completeGraph.RepoRoot)
+
+		// If there is an error, we can ignore it, since turbo.json config is not required in the workspace.
+		if workspaceTurboJSON, err := fs.ReadTurboConfig(workspaceConfigPath); err == nil {
+			// TODO(mehulkar): Enable extending from more than one workspace.
+			if len(workspaceTurboJSON.Extends) > 1 {
+				return nil, fmt.Errorf(
+					"You can only extend from the root workspace. \"%s\" extends from %v",
+					pkg.Name,
+					workspaceTurboJSON.Extends,
+				)
+			}
+
+			// TODO(mehulkar):
+			// 		Pipeline.GetTask allows searching with a taskID (e.g. `package#task`).
+			// 		But we do not want to allow this, except if we're in the root workspace.
+			workspaceDefinition, err := workspaceTurboJSON.Pipeline.GetTask(taskID, taskName)
+
+			if err == nil {
+				taskDefinitions = append(taskDefinitions, *workspaceDefinition)
+			}
+
+			// If there is no Extends key, we are either in a workspace that didn't
+			// extend from anything, or in a single package repo, where the workspace is the root.
+			if len(workspaceTurboJSON.Extends) == 0 {
+				// For the single package case, we can just return the workspaceDefinition
+				// (It should be the same as the rootTaskDefinition)
+				if pkg.PackageJSONPath == "package.json" {
```

Comment:
Is there a better condition to check here? Can we avoid getting to this point in the single package case?